tests/utils/node/nodes.go

```diff
@@ -0,0 +1,177 @@
+// init operation failed.
+func (nodes Nodes) Init(ctx context.Context) (err error) {
+	initErrors := make(chan error)
+	for _, node := range nodes {
+		go func(node Node) {
+			err := node.Init(ctx) // takes 2 seconds
+			if err != nil {
+				err = fmt.Errorf("node %s failed to initialise: %w", node, err)
+			}
+			initErrors <- err
+		}(node)
+	}
+
+	for range nodes {
+		initErr := <-initErrors
+		if err == nil && initErr != nil {
+			err = initErr
+		}
+	}
+
+	return err
+}
+
+// Start starts all the nodes and returns the number of started nodes
+// and an eventual start error. The started number should be used by
+// the caller to wait for `started` errors coming from the wait error
+// channel. All the nodes are stopped when the context is canceled,
+// and `started` errors will be sent in the waitErr channel.
+func (nodes Nodes) Start(ctx context.Context) (
+	runtimeErrors []<-chan error, startErr error) {
+	runtimeErrors = make([]<-chan error, 0, len(nodes))
+	for _, node := range nodes {
+		runtimeError, err := node.Start(ctx)
+		if err != nil {
+			return runtimeErrors, fmt.Errorf("node with index %d: %w",
+				*node.index, err)
+		}
+
+		runtimeErrors = append(runtimeErrors, runtimeError)
+	}
+
+	for _, node := range nodes {
+		port := node.RPCPort()
+		err := waitForNode(ctx, port)
+		if err != nil {
+			return runtimeErrors, fmt.Errorf("node with index %d: %w", *node.index, err)
+		}
+	}
+
+	return runtimeErrors, nil
+}
+
+// InitAndStartTest is a test helper method to initialise and start nodes,
+// as well as registering appriopriate test handlers.
+// If any node fails to initialise or start, cleanup is done and the test
+// is instantly failed.
+// If any node crashes at runtime, all other nodes are shutdown,
+// cleanup is done and the passed argument `signalTestToStop`
+// is called to signal to the main test goroutine to stop.
+func (nodes Nodes) InitAndStartTest(ctx context.Context, t *testing.T,
+	signalTestToStop context.CancelFunc) {
+	t.Helper()
+
+	err := nodes.Init(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	nodesCtx, nodesCancel := context.WithCancel(ctx)
+	runtimeErrors := runtime.NewErrorsFanIn()
+
+	for _, node := range nodes {
+		runtimeError, err := node.Start(nodesCtx) // takes little time
+		if err == nil {
+			runtimeErrors.Add(node.String(), runtimeError)
+			continue
+		}
+
+		t.Errorf("Node %s failed to start: %s", node, err)
+
+		stopNodes(t, nodesCancel, runtimeErrors)
+		t.FailNow()
+	}
+
+	// this is run sequentially since all nodes start almost at the same time
+	// so waiting for one node will also wait for all the others.
+	// You can see this since the test logs out that all the nodes are ready
+	// at the same time.
+	for _, node := range nodes {
+		err := waitForNode(ctx, node.RPCPort())
+		if err == nil {
+			t.Logf("Node %s is ready", node)
+			continue
+		}
+
+		t.Errorf("Node %s failed to be ready: %s", node, err)
+		stopNodes(t, nodesCancel, runtimeErrors)
+		t.FailNow()
+	}
```

Comment:
why not just for the nodes in parallel and use a waitgroup?

Reply:
See the comment

```go
	// this is run sequentially since all nodes start almost at the same time
	// so waiting for one node will also wait for all the others.
	// You can see this since the test logs out that all the nodes are ready
	// at the same time.
```

We could do it in parallel, but it adds quite a bit of complexity for no practical advantage.